### PR TITLE
[SPARK-3586][streaming]Support nested directories in Spark Streaming

### PR DESCRIPTION
**Pull request by zsxwing**
Closes #2765

### CODE REVIEW

**Comment by GordonWang** [comment]
@zsxwing Do you have any further plan about this PR ? When is it going to be merged?

Thanks.

**Comment by zsxwing** [comment]
> @zsxwing Do you have any further plan about this PR ? When is it going to be merged?
> 
> Thanks.

@tdas will look at this one soon. I think this one should be able to deliver in 1.5.0.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 1]
```diff
@@ -179,10 +187,55 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
...
+        if (status.isDir) {
+          // Note: A user may set depth = Int.MaxValue to search all nested directories.
+          if (depth > path.depth() - rootDirectoryDepth) {
+            if (lastFoundDirs.contains(path)) {
+              if (status.getModificationTime > modTimeIgnoreThreshold) {
+                fs.listStatus(path).foreach(searchFilesRecursively(_, files))
+              }
+            } else {
+              lastFoundDirs += path
+              fs.listStatus(path).foreach(searchFilesRecursively(_, files))
```
Comment (huitseeker) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: This (and the other recursive call on l.203 above) may blow up the call stack in very deep hierarchies (think Maildir++, etc). I realise there's a depth limit set, but users may precisely want to set it extremely high for some of those cases.

How about making the exploration using a @tailrec auxiliary function, with a queue of yet-to-be-explored directories in arguments ?

**Comment by andrewor14** [comment]
@tdas @zsxwing any updates on this? Is this targeted for 1.6 now?

**Comment by gprivitera** [comment]
@tdas : any update on this PR? Is there anything that should be done yet?

**Comment by gprivitera** [comment]
@zsxwing : your solution is not going to work on S3, since `org.apache.hadoop.fs.FileStatus.getModificationTime()` always returns `0L` on directories in S3.

**Comment by tdas** [comment]
Good point!

**Comment by srowen** [comment]
@zsxwing is this stalled now? It doesn't look like this will proceed.

**Comment by zsxwing** [comment]
Let me just close it. Maybe visit later.

**Comment by stshruthi** [comment]
Is there a plan to support nested directory streaming with latest versions of spark?

**Pull request closed without merging**